### ISSUE TITLE
Strange training perplexity

### ISSUE THREAD

**Issue by wsnooker** [OpenNMT/OpenNMT#122]
Hi, I found some strange perplexity jumping between epochs when training in my own dataset.

At epoch 1 end : train data perplexity is 14.37, and switches to 8.29 in the beginning of epoch 2.
Another question: why does the validation perplexity much lower than train perplexity at every epoch end. 

> Epoch 1 ; Iteration 74000/77776 ; Learning rate 1.0000 ; Source tokens/s 1270 ; Perplexity 14.64
> Epoch 1 ; Iteration 75000/77776 ; Learning rate 1.0000 ; Source tokens/s 1270 ; Perplexity 14.55
> Epoch 1 ; Iteration 76000/77776 ; Learning rate 1.0000 ; Source tokens/s 1270 ; Perplexity 14.46
> Epoch 1 ; Iteration 77000/77776 ; Learning rate 1.0000 ; Source tokens/s 1270 ; Perplexity 14.37
> Validation perplexity: 7.8701437591397
> Saving checkpoint to 'model-qrt/model-diff-0.5-0.8-10m_epoch1_7.87.t7'...
> 
> Epoch 2 ; Iteration 1000/77776 ; Learning rate 1.0000 ; Source tokens/s 1274 ; Perplexity 8.29
> Epoch 2 ; Iteration 2000/77776 ; Learning rate 1.0000 ; Source tokens/s 1269 ; Perplexity 8.29
> Epoch 2 ; Iteration 3000/77776 ; Learning rate 1.0000 ; Source tokens/s 1270 ; Perplexity 8.29
> Epoch 2 ; Iteration 4000/77776 ; Learning rate 1.0000 ; Source tokens/s 1268 ; Perplexity 8.28

**Comment by guillaumekln** [comment]
Perplexity is currently reported as an average over the epoch. So at 77000/77776, the actual perplexity is closer to 8.29 but it reports the average perplexity from iteration 1 to 77000.

Now that I think about it, I don't think that is a good behavior as it is somewhat confusing. I think the original goal was to show a consistent decrease in perplexity for the user.

**Comment by wsnooker** [comment]
Yes, this is confused.  As a general purpose, it should not be the average perplexity since every batch perplexity may be used to think about the convergence .

And  why does the validation perplexity much lower than train perplexity at every epoch end ?
Such as, training perplexity 8.29 while validation perplexity is 7.87. From my experience, validation perplexity usually is above training perplexity.

**Comment by guillaumekln** [comment]
I agree, we will certainly drop averaged metrics.

> And why does the validation perplexity much lower than train perplexity at every epoch end ?

Dropout is not applied during evaluation on the validation set. That may explain the lower perplexity.

**Comment by wsnooker** [comment]
From my point, the dropout is only used between different rnn layers, such as `rnn_layer1 -> dropout -> rnn_layer2`.  In my model there is only just one rnn layer, so dropout may not be used.  Are there other places that dropout is used ?  What about output of wordEmbedding ? Such as, `text -> wordEmbedding -> dropout -> rnn` .

**Comment by guillaumekln** [comment]
It is also applied on the output of the attention layer.

**Comment by wsnooker** [comment]
ok, thank you.

**Comment by guillaumekln** [comment]
Actual perplexity reporting during training will be part of #100.
